singer/src/main/java/com/pinterest/singer/writer/KafkaWriter.java

```diff
@@ -221,19 +222,85 @@ private void includeAuditMessageInBatch(List<ProducerRecord<byte[], byte[]>> mes
+                                                     List<LogMessage> logMessages) throws Exception {
+    LOG.info("Collate " + logMessages.size() + " messages");
+    List<List<LogMessage>> messageBuckets = new ArrayList<>();
+
+    try {
+      List<PartitionInfo> validPartitions = partitions;
+      if (skipNoLeaderPartitions) {
+        validPartitions = new ArrayList<>();
+        for (PartitionInfo partitionInfo : partitions) {
+          // If there is no leader, the id value is -1
+          // github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/PartitionInfo.java
+          if (partitionInfo.leader().id() >= 0) {
+            validPartitions.add(partitionInfo);
+          }
+        }
+      }
+      for (int i = 0; i < validPartitions.size(); i++) {
+        messageBuckets.add(new ArrayList<>());
+      }
+      for (LogMessage msg : logMessages) {
+        byte[] key = null;
+        if (msg.isSetKey()) {
+          key = msg.getKey();
+        }
+        int partitionId = partitioner.partition(key, validPartitions);
+        if (skipNoLeaderPartitions) {
+          partitionId = validPartitions.get(partitionId).partition();
+        }
+        messageBuckets.get(partitionId).add(msg);
+      }
+    } catch (Exception e) {
+      LOG.error("Failed in message collation for topic {}, partitioner {}",
+          topic, partitioner.getClass().getName(), e);
+      throw new LogStreamWriterException(e.toString());
+    }
+    return messageBuckets;
+  }
+
+  public List<List<ProducerRecord<byte[], byte[]>>> createProducerRecordFromLogMessage(
+      List<List<LogMessage>> messageBuckets, String topic){
+    List<List<ProducerRecord<byte[], byte[]>>> recordBuckets = new ArrayList<>();
+
+    ProducerRecord<byte[], byte[]> keyedMessage;
+    for(int partitionId = 0; partitionId < messageBuckets.size(); partitionId++) {
+      List<ProducerRecord<byte[], byte[]>> records = new ArrayList<>();
+      for(LogMessage msg: messageBuckets.get(partitionId)) {
+        byte[] key = null;
+        if (msg.isSetKey()) {
+          key = msg.getKey();
+        }
+        keyedMessage = new ProducerRecord<>(topic, partitionId, key, msg.getMessage());
+        Headers headers = keyedMessage.headers();
+        HeaderProvider.addEnvHeaders(headers);
+        records.add(keyedMessage);
+      }
+      recordBuckets.add(records);
+    }
+    return recordBuckets;
+  }
+
   @Override
   public void writeLogMessages(List<LogMessage> logMessages) throws LogStreamWriterException {
     Set<Future<KafkaWritingTaskResult>> resultSet = new HashSet<>();
     KafkaProducer<byte[], byte[]> producer = KafkaProducerManager.getProducer(producerConfig);
     Preconditions.checkNotNull(producer);
-
     if (producerConfig.isTransactionEnabled()) {
       producer.beginTransaction();
     }
     try {
       List<PartitionInfo> partitions = producer.partitionsFor(topic);
+      /*
```

Comment:
Remove commented out code.

Reply:
Why do we need to break the previous methods into two methods?  Could header information be attached to the List<List<ProducerRecord>> generated from messageCollation() method?